templates/base/client/.rtag/Forms.tsx.hbs

```diff
@@ -0,0 +1,296 @@
+import { RtagContext } from "./context";
+import { Response } from "./base";
+import {
+  {{#each methods}}
+  {{makeRequestName @key}},
+  {{/each}}
+  {{#each types}}
+  {{@key}},
+  {{/each}}
+} from "./types";
+
+type MethodProps<T> = {
+  method: string;
+  submit: (value: T) => Promise<Response>;
+  initialize: () => T;
+  children: (value: T, update: (value: T) => void) => JSX.Element;
+};
+type BaseProps<T> = {
+  value: T;
+  update: (value: T) => void;
+};
+type ContainerProps<T> = {
+  initialVal: T;
+  children: (value: T, update: (value: T) => void) => JSX.Element;
+};
+type ArrayProps<T> = BaseProps<T[]> & ContainerProps<T>;
+type OptionalProps<T> = BaseProps<T | undefined> & ContainerProps<T>;
+type EnumProps = BaseProps<number> & { enum: object };
+
+export function MethodForm<T>(props: MethodProps<T>) {
+  const [value, setValue] = useState<T>(props.initialize());
+  console.log("MethodForm", value);
+  return (
+    <div className="method-form">
+      <h3>{props.method}</h3>
+      {props.children(value, setValue)}
+      <button
+        onClick={async () => {
+          const res = await props.submit(value);
+          if (res.type === "error") {
+            toast.error(res.error);
+          }
+          setValue(props.initialize());
+        }}
+      >
+        Submit
+      </button>
+    </div>
+  );
+}
+
+function ArrayInput<T>(props: ArrayProps<T>) {
+  function swapArgs(i1: number, i2: number) {
+    [props.value[i1], props.value[i2]] = [props.value[i2], props.value[i1]];
+    props.update(props.value);
+  }
+  function deleteArg(i: number) {
+    props.value.splice(i, 1);
+    props.update(props.value);
+  }
+  return (
+    <div className="array-input">
+      {props.value.map((item, i) => (
+        <div key={i} className="array-item">
+          {props.children(item, (val) => {
+            props.update(Object.assign(props.value, { [i]: val }));
+          })}
+          <button disabled={i === 0} onClick={() => swapArgs(i, i - 1)}>
+            &#8593;
+          </button>
+          <button disabled={i === props.value.length - 1} onClick={() => swapArgs(i, i + 1)}>
+            &#8595;
+          </button>
+          <button onClick={() => deleteArg(i)}>x</button>
+        </div>
+      ))}
+      <button onClick={() => props.update(props.value.concat(props.initialVal))}>Add</button>
+    </div>
+  );
+}
+
+function OptionalInput<T>(props: OptionalProps<T>) {
+  return (
+    <div className="optional-input">
+      {props.value !== undefined
+        ? props.children(props.value, (val) => {
+            props.update(val);
+          })
+        : ""}
+      <button onClick={() => (props.value === undefined ? props.update(props.initialVal) : props.update(undefined))}>
+        Toggle
+      </button>
+    </div>
+  );
+}
+
+function EnumInput(props: EnumProps) {
+  return (
+    <div className="enum-input">
+      <select value={props.value} onChange={(e) => props.update(Number(e.target.value))}>
```

Comment:
this feedback i have about functions defined outside useCallback applies to a lot of places